src/sctools/bam.py

```diff
@@ -0,0 +1,125 @@
+import pysam
+
+"""
+unlike fastq and gtf which lack flexible iterators, the pysam wrapper provides an excellent iterator 
+over sam records that would be difficult to improve upon in pure python. 
+
+This module provides a few additional methods that extend the functionality of pysam
+"""
+
+
+class SubsetAlignments:
+    """Wrapper for pysam/htslib that enables non-standard filtering of alignments in a bamfile
+
+    :method indices_by_chromosome: extract reads specific to a chromosome
+    """
+
+    def __init__(self, alignment_file, open_mode=None):
+        """
+        :param str alignment_file: sam or bam file.
+        :param str open_mode: optional, mode to read file. Will be autodetected by file type if
+          the file contains the correct suffix for its type.
+        """
+
+        if open_mode is None:
+            if alignment_file.endswith('.bam'):
+                open_mode = 'rb'
+            elif alignment_file.endswith('.sam'):
+                open_mode = 'r'
+            else:
+                raise ValueError('could not autodetect file type for alignment_file %s '
+                                 '(detectible suffixes: .sam, .bam)' % alignment_file)
+        self._file = alignment_file
+        self._open_mode = open_mode
+
+    def indices_by_chromosome(self, n_specific, chromosome, include_other=0):
```

Comment:
Maybe call it include_unaligned instead of include_other. Would it make sense to interpret -1 to mean include all unaligned reads? Is that a useful thing to be able to do?

Reply:
Oh I see, this is not just unaligned reads but any read that does not match the chromosome we asked for.

Reply:
i'll find a way to clarify; sounds like it produced some confusion.

Reply:
Adjusted docstring: 

"""Return the list of first n_specific indices of reads aligned to selected chromosome.

        If desired, will also return non-specific indices in a second list (can serve as negative
        control reads).

        :param int n_specific: number of aligned reads to return indices for
        :param str chromosome: only reads from this chromosome are considered valid
        :param int include_other: optional, (default=0), the number of reads to include that are
          NOT aligned to chromosome (could be aligned or unaligned read)
        :return [int]: list of indices to reads aligning to chromosome
        :return [int]: list of indices to reads NOT aligning to chromosome, only returned if
          include_other is not 0.
        """